### ISSUE TITLE
dropdown never closes - security policies may prevent inline styles

### ISSUE THREAD

**Issue by KristerV** [maxmarcon/live_select#16]
so i tried live_select again and the dropdown does not close indeed.

phoenix 1.7, live_view 0.18, but my empty test project had the same stuff and no problem there.

i downloaded the lib and connected the local version to my app to debug.

it seems when it comes down to it the `@hide_dropdown` variable is set correctly, but the `display: none;` doesn't end up in the HTML.

Here's an image where i print the var out. this is with initial options only, no interaction.

![image](https://user-images.githubusercontent.com/2750622/216336539-0bb30504-5f64-43f8-a11c-dd756556bed3.png)

another interesting thing i noticed was that the initial HTML coming from the server does in fact have `style="display: none;"` included, but then the websocket gets two updates immediately and the second one has an empty class and style for this element. which is another confusing part, because the class does end up with appropriate classes so why doesn't style?

this is my thought process so far. i have no idea what it is in my app that is causing this. i thought perhaps you've got ideas where to look?

edit: i'm placing random values in places and it seems like none of the style attributes in the whole component get shown on page. not sure what this may or may not mean.

edit2: experimenting with raw style values it seems that it's not the LiveView websocket update that empties the style. so maybe javascript, but my project has barely any JS at all. new suspect: a library.

**Comment by KristerV** [comment]
okay i've confirmed the bug is not to do with this library, but my own app somehow. no styles anywhere get saved.

**Comment by KristerV** [comment]
hmm.. so i've found the solution to my issue. and it's about security, so I'll open this again and let you close it so i can be sure you've read it at least.

i tracked the issue all the way down to my third commit in the project. that's when i installed Sobelow and it told me that my Content-Security-Policy wasn't set properly. So I googled areound and set it to the recommended default: 
```
plug :put_secure_browser_headers, %{"content-security-policy" => "default-src 'self'"}
```
clearly i don't know much about it since this is what disallows inline stuff, like the style attribute. now I added 'unsafe-inline' to it and live_select works again. mind you - nothing else was broken in the app so i guess no need to use it really.

i don't know how dangerous unsafe-inline is honestly, but you may want to consider not using style for the functionality. and seeing this did make me wonder why you didn't just use an `if` case?

**Comment by maxmarcon** [comment]
thanks @KristerV for sharing this. It's very interesting.

I wasn't aware that security policies might forbid inline styles - this is really good to now. Damn security! 😆  

This needs to be changed.

> why you didn't just use an if case?

Good question! 😄 Looking at the code for the dropdown, yeah, an `if` seems to be the obvious solution. To be honest, I don't remember why I chose to use inline styles instead. However, I'm pretty sure I must have had a reason (although the reason might not be relevant anymore). Maybe something wasn't working with `if` so I tried the inline style and I never revisited the decision.

I'm using linline styles also [here](https://github.com/maxmarcon/live_select/blob/main/lib/live_select/component.html.heex#L52) instead of a hidden input because of bug in LV (now fixed, but maybe not released yet, and anyway the library should work with older versions as well). This might be trickier to change, though.

I will try to remove inline styles ASAP. Thanks again!

**Comment by KristerV** [comment]
I would imagine one benefit of inline styles is that if the options don't change then opening and closing the dropdown will just have much less network traffic.

I think doing the same with classes is still fine. If it works. Tailwind doesn't play to well with overwrites.

**Comment by maxmarcon** [comment]
Btw, I thought this should be an issue for LV as well, because it uses inline styles too, for example to make elements disappear with the JS.hide command.

However, I tested it and JS.hide works even with your restrictive security header.

I suspect they must be doing it this way on the client side: `document.querySelector("div").style.display = "none";`
as documented here:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src

**Comment by maxmarcon** [comment]
BTW: I tested using `if` instead of the inline style. It doesn't work: you can't select any entry. So there was a reason :)

I suspect it hast to do with events firing for the selected element when the element in the dropdown is no longer there (because of the `if`)

I will inviestigate

**Comment by KristerV** [comment]
> I suspect they must be doing it this way on the client side:

and so the rabbit hole deepens. interesting stuff.

> events firing for the selected element when the element in the dropdown is no longer there 

not sure what you mean here. you click on an item in the dropdown, then the event fires, handler does stuff and changes state, then dropdown disappears. is this not how it goes?

**Comment by maxmarcon** [comment]
>not sure what you mean here. you click on an item in the dropdown, then the event fires, handler does stuff and changes state, then dropdown disappears. is this not how it goes?

yes my speculation turned out to be wrong. It wasn't working because of some JS thingy that had to be adpated (event handlers trying to attach to nonexistent elements).

Now it works. However, I'm using inline styles in a few other places and I will have to change those too (setting position: relative for the "x" clear button and hiding the text input that contains the value of the selection)

I can do it via JS but it sucks, because on page load the user will see the element that's supposed to be hidden until JS kicks in. Same happesn when LV patches the DOM.

I see no way around this problem other than adding some CSS for the component (which I wanted to avoid).

I would be nice if there was a way to provide scoped CSS for components, but I couldn't find anything that does that so far (apart from maybe the Surface library)

**Comment by KristerV** [comment]
Scoped css? Tailwind itself is scoped. Can't you use that?

Or if you need custom classes then prefixing them with live_select_whatever is an option.

**Comment by maxmarcon** [comment]
> Scoped css? Tailwind itself is scoped. Can't you use that?

The component needs to work also without tailwind. That's why I was using inline styles instead of tailwind classes :)

> Or if you need custom classes then prefixing them with live_select_whatever is an option.

Yes, of course. But the big question is how to distribute these classes. I wouldn't want to ask the user to add extra CSS to their application manually - more installation hassle. It would be nice if LV provided a way for components to bring their own scoped CSS - something you can define as part of the code of your components. I'm pretty sure all modern JS frontends have this feature, and in the LV world Surface has it too.

**Comment by KristerV** [comment]
Oh yes, good point.

How does surface do it?

I'm traveling now, can't check anything for a while.

**Comment by maxmarcon** [comment]
Yeah safe travels and don't worry! If you come up with some idea, let me know. I'll look deeper into this when I find some time.

Thanks for your interest.

**Comment by maxmarcon** [comment]
Hi @KristerV . I believe I fixed the problem. If you have time, you can try your application with strict security policies enabled against the main branch. Let me know if it works.

**Comment by KristerV** [comment]
tested and works! good job.

i went looking around the diff to see how you've worked around this and overall looks great, just couldn't quite figure out why you're putting the lear button styles each update?

![image](https://user-images.githubusercontent.com/2750622/217188750-bb30e8f1-4cdd-4ba6-a09b-9f86353dd63c.png)

**Comment by maxmarcon** [comment]
thanks for testing!

> isn't once (on mount) enough?

Unfortunately not. On mount there could be no selection yet, and so no clear button to style. So I have to look for a button and apply the styles on each DOM patch event.

**Comment by maxmarcon** [comment]
closed with d6a63b2f4775f8e494e2938dadbe1e9404b414f0
